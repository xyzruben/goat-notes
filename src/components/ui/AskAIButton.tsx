"use client";

import React from "react";
import { User } from "@supabase/supabase-js"
import { Button } from "@/components/ui/button"
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from "@/components/ui/dialog"
import { Fragment, useRef, useState, useTransition } from "react";
import { useRouter } from "next/navigation";
import { Textarea } from "./textarea";
import { ArrowUpIcon } from "lucide-react";
import { askAIAboutNotesAction } from "@/actions/notes";
import '@/app/styles/ai-response.css';
import DOMPurify from 'isomorphic-dompurify';


type Props = {
    user: User | null; 
};

function AskAIButton({user}: Props) {
    const router = useRouter();

    const [isPending] = useTransition();


    const [open, setOpen] = useState(false);
    const [questionText, setQuestionText] = useState("");
    const [questions, setQuestions] = useState<string[]>([]);
    const [responses, setResponses] = useState<string[]>([]);

    const handleOnOpenChange = (isOpen: boolean) => {
      if (!user) {
        router.push("/login")
      } else {
        if (isOpen) {
          setQuestionText("");
          setQuestions([]);
          setResponses([]);

        }
        setOpen(isOpen)
      }
    };

    const textareaRef = useRef<HTMLTextAreaElement>(null);
    const contentRef = useRef<HTMLDivElement>(null);

    const handleInput = () => {
      const textarea = textareaRef.current;
      if (!textarea) return;

      textarea.style.height = "auto";
      textarea.style.height = `${textarea.scrollHeight}px`;
    }

    const handleClickInput = () => {
      textareaRef.current?.focus();
    }

    const handleSubmit = () => {
      if (!questionText.trim()) return;

      const newQuestions = [...questions, questionText];
      setQuestions(newQuestions);
      setQuestionText("");
      setTimeout(scrollToBottom, 100);

      (async () => {
        const response = await askAIAboutNotesAction(newQuestions, responses);
        setResponses(prev => [...prev, response]);
        setTimeout(scrollToBottom, 100);
      })();
    };

    const scrollToBottom = () => {
      contentRef.current?.scrollTo({
        top: contentRef.current.scrollHeight,
        behavior: "smooth",
      })
    }

    const handleKeyDown = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
      if (e.key==="Enter" && !e.shiftKey) {
        e.preventDefault();
        handleSubmit();
      }

    }

  return <Dialog open={open} onOpenChange={handleOnOpenChange}>
  <DialogTrigger asChild>
    <Button variant="secondary">Ask AI</Button>
  </DialogTrigger>
  <DialogContent 
     className="custom-scrollbar flex h-[85vh] max-w-4xl flex-col overflow-y-auto" 
     ref={contentRef}
     >
    <DialogHeader>
      <DialogTitle>Ask AI About Your Notes</DialogTitle>
      <DialogDescription>
        Our AI can answer questions about all of your notes
      </DialogDescription>
    </DialogHeader>
    <div className="mt-4 flex flex-col gap-8">
      {questions.map((question, index) => (
        <Fragment key={index}>
          <p className="bg-muted text-muted-foreground ml-auto max-w-[60%] rounded-md px-2 py-1 text-sm">
            {question}
          </p>
          {
            responses[index] && (
              <p
              className="bot-response text-muted-foreground text-sm"
              dangerouslySetInnerHTML={{
                __html: DOMPurify.sanitize(responses[index], {
                  ALLOWED_TAGS: ['p', 'strong', 'em', 'ul', 'ol', 'li', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'br'],
                  ALLOWED_ATTR: []
                })
              }}

               />
            )
          }
        </Fragment>
      ))}
      {isPending && <p className="animate-pulse text-sm">Thinking....</p>}
    </div>
    <div 
    className="mt-auto flex cursor-text flex-col rounded-lg border p-4" 
    onClick={handleClickInput}
    >
      <Textarea 
      ref={textareaRef} 
      placeholder="Ask me anything about your notes..." 
      className="resize-none rounded-none border-none bg-transparent p-0 shadow-none 
      placeholder:text-muted-foreground focus-visible:ring-0 focus-visible:ring-offset-0" 
      style={{
        minHeight: "0",
        lineHeight: "normal",
      }}
      rows={1}
      onInput={handleInput}
      onKeyDown={handleKeyDown}
      value={questionText}
      onChange={(e) => setQuestionText(e.target.value)}
       />

    </div>
    <Button className="ml-auto size-8 rounded-full" onClick={handleSubmit}>
      <ArrowUpIcon className="text-background" />
    </Button>
    

  </DialogContent>
</Dialog>
}

export default AskAIButton;