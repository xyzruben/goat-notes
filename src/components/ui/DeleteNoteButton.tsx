"use client"

import {
  AlertDialog,
  AlertDialogAction,
  AlertDialogCancel,
  AlertDialogContent,
  AlertDialogDescription,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogTitle,
  AlertDialogTrigger,
} from "@/components/ui/alert-dialog"
import { Button } from "./button";
import { Loader2, Trash2 } from "lucide-react";
import { useTransition } from "react";
import { useToast } from "@/hooks/use-toast";
import { useRouter, useSearchParams } from "next/navigation";
import { deleteNoteAction } from "@/actions/notes";


type Props = {
    noteId: string;
    deleteNoteLocally: (noteId: string) => void;
};

function DeleteNoteButton({ noteId, deleteNoteLocally }: Props) {

  const router = useRouter();
  const {toast} = useToast();
  const noteIdParam = useSearchParams().get("noteId") || ""

  const [isPending, startTransition] = useTransition();

  const handleDeleteNote = () => {
    startTransition(async () => {
      const {errorMessage} = await deleteNoteAction(noteId);

      if (!errorMessage) {
        toast({
          title: "Note deleted",
          description: "You have successfully deleted the Note",
          variant: "success"
        })
        deleteNoteLocally(noteId);

        if (noteId === noteIdParam) {
          router.replace("/")
        }

      } else {
        toast({
          title: "Error",
          description: errorMessage,
          variant: "destructive"
        })
      }
      
    })
  }

  return (
    <AlertDialog>
  <AlertDialogTrigger asChild>
    <Button 
    className="absolute right-2 top-1/2 -translate-y-1/2 size-7 p-0 opacity-0 group-hover/item:opacity-100 [&_svg]:size-4" 
    variant="ghost"
    ><Trash2 />
    </Button>
  </AlertDialogTrigger>
  <AlertDialogContent>
    <AlertDialogHeader>
      <AlertDialogTitle>Are you sure you want to delete this note?</AlertDialogTitle>
      <AlertDialogDescription>
        This action cannot be undone. This will permanently delete your Note
        and remove your data from our servers.
      </AlertDialogDescription>
    </AlertDialogHeader>
    <AlertDialogFooter>
      <AlertDialogCancel>Cancel</AlertDialogCancel>
      <AlertDialogAction onClick={handleDeleteNote} 
      className="bg-destructive text-destructive-foreground hover:bg-destructive/90 w-24"
      >
        {isPending ? <Loader2 className="animate-spin" /> : "Delete"}
      </AlertDialogAction>
    </AlertDialogFooter>
  </AlertDialogContent>
</AlertDialog>

  )
  
}

export default DeleteNoteButton;