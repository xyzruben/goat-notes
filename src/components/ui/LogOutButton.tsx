"use client";

import React from "react";
import { Loader2 } from "lucide-react"
import { Button } from "./button"
import { useState } from "react";
import { useToast } from "@/hooks/use-toast";
import { useRouter } from "next/navigation";
import { logOutAction } from "@/actions/users";
import { supabase } from "../../../__mocks__/supabase";

function LogOutButton() {

    const {toast} = useToast();
    const router = useRouter();

  const [loading, setLoading] = useState(false);

  const handleLogOut = async () => {
    setLoading(true)
    const { errorMessage } = await logOutAction();

    if (!errorMessage) {
        toast({
            title: "Logged out",
            description: "You have been successfully Logged out",
            variant: "success",
        });
        router.push("/");
    } else {
      toast({
        title: "Error",
        description: errorMessage,
        variant: "destructive",
      })
    }

    setLoading(false)
  }

  return (
    <Button 
    variant="outline"
    onClick={handleLogOut}
    disabled={loading}
    className="width-24">
        {loading ? <Loader2 className="animate-spin" /> : "Log Out"}
    </Button>
  
);
}

export default LogOutButton;