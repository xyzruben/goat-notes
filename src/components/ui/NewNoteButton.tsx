"use client";

import { User } from "@supabase/supabase-js";
import { Button } from "./button";
import { Loader2 } from "lucide-react";
import { useState } from "react";
import { useRouter } from "next/navigation";
import { v4 as uuidv4 } from "uuid";
import { useToast } from "@/hooks/use-toast";
import { createNoteAction } from "@/actions/notes";

type Props = {
    user: User | null;
};

function NewNoteButton({ user }: Props) {
    const router = useRouter();

    const {toast} = useToast()

    const [loading, setLoading] = useState(false);

    const handleClickNewNoteButton = async () => {
      if (!user) {
        router.push("/login")
      } else {
        setLoading(true)

        const uuid = uuidv4()
        await createNoteAction(uuid)
        router.push(`/?noteId=${uuid}`)

        toast({
          title: "New Note Created",
          description: "You have created a new note",
          variant: "success"
        });
        setLoading(false)
      }
    }

  return(
  <Button 
  onClick={handleClickNewNoteButton} 
  variant="secondary" 
  className="w-24" 
  disabled={loading}

  >
    {loading ? <Loader2 className="animate-spin" /> : "New Note"}
  </Button>
  );
}

export default NewNoteButton;