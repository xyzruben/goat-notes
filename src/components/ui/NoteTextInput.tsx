"use client";

import { useSearchParams } from "next/navigation";

type Props = {
    noteId: string;
    startingNoteText: string;
}

function NoteTextInput({ noteId, startingNoteText}: Props) {
    const noteIdParam = useSearchParams().get("noteId") || "";
    const {noteId, setNoteId} = useNote();
  return (
    <div>NoteTextInput</div>
  )
}

export default NoteTextInput;