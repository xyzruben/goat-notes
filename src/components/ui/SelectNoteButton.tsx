'use client';

import useNote from "@/hooks/useNote";
import { Note } from "@prisma/client";
import { useSearchParams } from "next/navigation";
import { useState } from "react";

type Props = {
    note: Note;
};

function SelectNoteButton({ note }: Props) {
  const noteId = useSearchParams().get("noteId") || "";

  const {noteText: selectedNoteText} = useNote();
  const [shouldUseGlobalNoteText, setShouldUseGlobalNoteText] = useState(false)
  const [localNoteText, setLocalNoteText] = useState(note.text)

  const blankNoteText = "EMPTY NOTE";
  let noteText = localNoteText || blankNoteText;
  if (shouldBeGlobalNoteText) {
    noteText = selectedNoteText || blankNoteText;
  }

  return (
    <div>SelectNoteButton</div>
  )
}

export default SelectNoteButton;