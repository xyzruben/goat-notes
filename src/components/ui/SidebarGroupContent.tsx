"use client";

import { Note } from "@prisma/client";
import { SidebarGroupContent as SidebarGroupContentShadCN, SidebarMenu } from "./sidebar";
import { SearchIcon } from "lucide-react";
import { Input } from "./input";
import { useEffect, useMemo, useState } from "react";
import Fuse from "fuse.js"

type Props = {
    notes: Note[];
};

function SidebarGroupContent( {notes}: Props ) {
  const [searchText, setSearchText] = useState("");
  const [localNotes, setLocalNotes] = useState(notes);

  useEffect(() => {
    setLocalNotes(notes)
  }, [notes])

  const fuse = useMemo(() => {
    return new Fuse(localNotes, {
      keys: ["text"],
      threshold: .4
    })
  }, [localNotes])

  const filteredNotes = searchText ? fuse.search(searchText).map(result => result.item) : localNotes;

  return (
    <SidebarGroupContentShadCN>
      <div className="relative flex items-center">
        <SearchIcon className="absolute left-2 size-4" />
        <Input
        className="bg-muted pl-8" 
        placeholder="Search your notes..."
        value={searchText} 
        onChange={(e) => setSearchText(e.target.value)}
        />
      </div>
      <SidebarMenu className="mt-4">

      </SidebarMenu>
    </SidebarGroupContentShadCN>
  )
}

export default SidebarGroupContent;