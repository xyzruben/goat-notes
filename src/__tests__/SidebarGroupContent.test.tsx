import React from 'react';
import { render } from '@testing-library/react';
import SidebarGroupContent from '@/components/ui/SidebarGroupContent';
import NoteProvider from '@/providers/NoteProvider';
import { SidebarProvider } from '@/components/ui/sidebar';

// Mock Next.js navigation
jest.mock('next/navigation', () => ({
  useSearchParams: () => ({
    get: jest.fn().mockReturnValue(''),
  }),
  useRouter: () => ({
    push: jest.fn(),
    refresh: jest.fn(),
  }),
}));

// Mock notes data with fixed dates for consistent snapshots
const mockNotes = [
  {
    id: '1',
    text: 'Test note 1',
    authorId: 'user1',
    createdAt: new Date('2025-06-20T10:00:00Z'),
    updatedAt: new Date('2025-06-20T10:00:00Z'),
  },
  {
    id: '2',
    text: 'Test note 2',
    authorId: 'user1',
    createdAt: new Date('2025-06-20T10:00:00Z'),
    updatedAt: new Date('2025-06-20T10:00:00Z'),
  },
];

const wrapper = ({ children }: { children: React.ReactNode }) => (
  <SidebarProvider>
    <NoteProvider>{children}</NoteProvider>
  </SidebarProvider>
);

describe('SidebarGroupContent', () => {
  it('renders without crashing', () => {
    const { container } = render(<SidebarGroupContent notes={mockNotes} />, { wrapper });
    expect(container).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<SidebarGroupContent notes={mockNotes} />, { wrapper });
    expect(container).toMatchSnapshot();
  });
}); 