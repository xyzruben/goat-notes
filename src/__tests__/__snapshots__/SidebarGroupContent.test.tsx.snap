// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SidebarGroupContent matches snapshot 1`] = `
<div>
  <div
    class="group/sidebar-wrapper has-data-[variant=inset]:bg-sidebar flex min-h-svh w-full"
    data-slot="sidebar-wrapper"
    style="--sidebar-width: 16rem; --sidebar-width-icon: 3rem;"
  >
    <div
      class="w-full text-sm"
      data-sidebar="group-content"
      data-slot="sidebar-group-content"
    >
      <div
        class="relative flex items-center"
      >
        <svg
          class="lucide lucide-search absolute left-2 size-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <path
            d="m21 21-4.3-4.3"
          />
        </svg>
        <input
          class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-muted pl-8"
          data-slot="input"
          placeholder="Search your notes..."
          value=""
        />
      </div>
      <ul
        class="flex w-full min-w-0 flex-col gap-1 mt-4"
        data-sidebar="menu"
        data-slot="sidebar-menu"
      >
        <li
          class="group/menu-item relative group/item"
          data-sidebar="menu-item"
          data-slot="sidebar-menu-item"
        >
          <a
            class="peer/menu-button flex w-full overflow-hidden rounded-md p-2 text-left outline-hidden ring-sidebar-ring transition-[width,height,padding] focus-visible:ring-2 active:bg-sidebar-accent active:text-sidebar-accent-foreground disabled:pointer-events-none disabled:opacity-50 group-has-data-[sidebar=menu-action]/menu-item:pr-8 aria-disabled:pointer-events-none aria-disabled:opacity-50 data-[active=true]:bg-sidebar-accent data-[active=true]:font-medium data-[active=true]:text-sidebar-accent-foreground data-[state=open]:hover:bg-sidebar-accent data-[state=open]:hover:text-sidebar-accent-foreground group-data-[collapsible=icon]:size-8! group-data-[collapsible=icon]:p-2! [&>span:last-child]:truncate [&>svg]:size-4 [&>svg]:shrink-0 hover:bg-sidebar-accent hover:text-sidebar-accent-foreground h-8 text-sm items-start gap-0 pr-12 false flex h-fit flex-col"
            data-active="false"
            data-sidebar="menu-button"
            data-size="default"
            data-slot="sidebar-menu-button"
            href="/?noteId=1"
          >
            <p
              class="w-full overflow-hidden truncate text-eclipsis whitespace-nowrap"
            >
              Test note 1
            </p>
            <p
              class="text-muted-foreground text-xs "
            >
              6/20/2025
            </p>
          </a>
          <button
            aria-controls="radix-«r6»"
            aria-expanded="false"
            aria-haspopup="dialog"
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 has-[>svg]:px-3 absolute right-2 top-1/2 -translate-y-1/2 size-7 p-0 opacity-0 group-hover/item:opacity-100 [&_svg]:size-4"
            data-slot="alert-dialog-trigger"
            data-state="closed"
            type="button"
          >
            <svg
              class="lucide lucide-trash2 lucide-trash-2"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 6h18"
              />
              <path
                d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
              />
              <path
                d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
              />
              <line
                x1="10"
                x2="10"
                y1="11"
                y2="17"
              />
              <line
                x1="14"
                x2="14"
                y1="11"
                y2="17"
              />
            </svg>
          </button>
        </li>
        <li
          class="group/menu-item relative group/item"
          data-sidebar="menu-item"
          data-slot="sidebar-menu-item"
        >
          <a
            class="peer/menu-button flex w-full overflow-hidden rounded-md p-2 text-left outline-hidden ring-sidebar-ring transition-[width,height,padding] focus-visible:ring-2 active:bg-sidebar-accent active:text-sidebar-accent-foreground disabled:pointer-events-none disabled:opacity-50 group-has-data-[sidebar=menu-action]/menu-item:pr-8 aria-disabled:pointer-events-none aria-disabled:opacity-50 data-[active=true]:bg-sidebar-accent data-[active=true]:font-medium data-[active=true]:text-sidebar-accent-foreground data-[state=open]:hover:bg-sidebar-accent data-[state=open]:hover:text-sidebar-accent-foreground group-data-[collapsible=icon]:size-8! group-data-[collapsible=icon]:p-2! [&>span:last-child]:truncate [&>svg]:size-4 [&>svg]:shrink-0 hover:bg-sidebar-accent hover:text-sidebar-accent-foreground h-8 text-sm items-start gap-0 pr-12 false flex h-fit flex-col"
            data-active="false"
            data-sidebar="menu-button"
            data-size="default"
            data-slot="sidebar-menu-button"
            href="/?noteId=2"
          >
            <p
              class="w-full overflow-hidden truncate text-eclipsis whitespace-nowrap"
            >
              Test note 2
            </p>
            <p
              class="text-muted-foreground text-xs "
            >
              6/20/2025
            </p>
          </a>
          <button
            aria-controls="radix-«r9»"
            aria-expanded="false"
            aria-haspopup="dialog"
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 has-[>svg]:px-3 absolute right-2 top-1/2 -translate-y-1/2 size-7 p-0 opacity-0 group-hover/item:opacity-100 [&_svg]:size-4"
            data-slot="alert-dialog-trigger"
            data-state="closed"
            type="button"
          >
            <svg
              class="lucide lucide-trash2 lucide-trash-2"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 6h18"
              />
              <path
                d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
              />
              <path
                d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
              />
              <line
                x1="10"
                x2="10"
                y1="11"
                y2="17"
              />
              <line
                x1="14"
                x2="14"
                y1="11"
                y2="17"
              />
            </svg>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
